### PR TITLE
Add ability to cancel Pants process

### PR DESCRIPTION
**Pull request by wisechengyi**
## Problem
Once Pants process is invoked from IntelliJ, there is no easy way to cancel it equivalent to Ctrl + C on command line, and user has to manually find the process on command line.

## Solution
Keep track of active Pants process per project under `PantsMakeBeforeRun` and add GUI interface to trigger process termination.

Interface looks like:
screen shot 2017-02-06 at 1 25 23 pm

### CODE REVIEW

File: src/com/twitter/intellij/pants/execution/PantsMakeBeforeRun.java [hunk 1]
```diff
@@ -121,6 +126,14 @@ else if (runConfiguration instanceof CommonProgramRunConfigurationParameters) {
     runManagerImpl.setBeforeRunTasks(runConfiguration, Collections.singletonList(pantsMakeTask), false);
   }
 
+  public static void terminatePantsProcess(Project project) {
+    Process process = runningPantsProcesses.get(project);
+    if (process != null) {
+      process.destroy();
+    }
+    runningPantsProcesses.remove(project);
```
Comment (baroquebobcat) [inline src/com/twitter/intellij/pants/execution/PantsMakeBeforeRun.java]: It'd be safer to use  [remove(Object, Object)](https://docs.oracle.com/javase/7/docs/api/java/util/concurrent/ConcurrentHashMap.html#remove(java.lang.Object,%20java.lang.Object))  here

File: src/com/twitter/intellij/pants/execution/PantsMakeBeforeRun.java [hunk 2]
```diff
@@ -264,7 +277,10 @@ public void onTextAvailable(ProcessEvent event, Key outputType) {
         output.add(event.getText());
       }
     });
+
+    runningPantsProcesses.put(currentProject, process);
     processHandler.runProcess();
+    runningPantsProcesses.remove(currentProject);
```
Comment (baroquebobcat) [inline src/com/twitter/intellij/pants/execution/PantsMakeBeforeRun.java]: same remove comment

File: src/com/twitter/intellij/pants/ui/PantsConsoleViewPanel.java [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+import org.jetbrains.annotations.NotNull;
+
+import javax.swing.JPanel;
+import javax.swing.JScrollPane;
+import javax.swing.tree.DefaultMutableTreeNode;
+import javax.swing.tree.DefaultTreeModel;
+import java.awt.BorderLayout;
+
+/**
+ * A JPanel consisting controll buttons on the left and a ConsoleView on the right displaying Pants output.
```
Comment (baroquebobcat) [inline src/com/twitter/intellij/pants/ui/PantsConsoleViewPanel.java]: nit: s/controll/of control/

File: src/com/twitter/intellij/pants/ui/PantsConsoleViewPanel.java [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+      Presentation presentation = getTemplatePresentation();
+      presentation.setIcon(AllIcons.Actions.Suspend);
+      presentation.setText(CommonBundle.getCloseButtonText());
+      presentation.setDescription(null);
+    }
+
+    @Override
+    public void actionPerformed(AnActionEvent e) {
+      PantsMakeBeforeRun.terminatePantsProcess(e.getProject());
+      PantsConsoleManager.getOrMakeNewConsole(e.getProject()).print("Pants process terminated.", ConsoleViewContentType.ERROR_OUTPUT);
```
Comment (baroquebobcat) [inline src/com/twitter/intellij/pants/ui/PantsConsoleViewPanel.java]: maybe move this string to `common/com/twitter/intellij/pants/PantsBundle.properties` ?

File: src/com/twitter/intellij/pants/ui/PantsConsoleViewPanel.java [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+
+  private class StopAction extends DumbAwareAction {
+    public StopAction() {
+      super(IdeBundle.message("action.stop"), null, AllIcons.Actions.Suspend);
+    }
+
+    @Override
+    public void actionPerformed(AnActionEvent e) {
+      PantsMakeBeforeRun.terminatePantsProcess(e.getProject());
+      PantsConsoleManager.getOrMakeNewConsole(e.getProject()).print("\nPants process terminated.", ConsoleViewContentType.ERROR_OUTPUT);
```
Comment (baroquebobcat) [inline src/com/twitter/intellij/pants/ui/PantsConsoleViewPanel.java]: same. It also feels like PantsProcessCancellationAction and StopAction could have this extracted to a common location. Unless there's a reason why they are slightly different.

File: src/com/twitter/intellij/pants/execution/PantsMakeBeforeRun.java (hunk 1 above)
Comment (wisechengyi) [inline src/com/twitter/intellij/pants/execution/PantsMakeBeforeRun.java]: corrected.

File: src/com/twitter/intellij/pants/execution/PantsMakeBeforeRun.java (hunk 2 above)
Comment (wisechengyi) [inline src/com/twitter/intellij/pants/execution/PantsMakeBeforeRun.java]: corrected.

File: src/com/twitter/intellij/pants/ui/PantsConsoleViewPanel.java (hunk 3 above)
Comment (wisechengyi) [inline src/com/twitter/intellij/pants/ui/PantsConsoleViewPanel.java]: corrected.

File: src/com/twitter/intellij/pants/ui/PantsConsoleViewPanel.java (hunk 4 above)
Comment (wisechengyi) [inline src/com/twitter/intellij/pants/ui/PantsConsoleViewPanel.java]: turns out there is something there already, so reused that.

File: src/com/twitter/intellij/pants/ui/PantsConsoleViewPanel.java (hunk 5 above)
Comment (wisechengyi) [inline src/com/twitter/intellij/pants/ui/PantsConsoleViewPanel.java]: `PantsProcessCancellationAction` removed for being redundant.

File: src/com/twitter/intellij/pants/ui/PantsConsoleViewPanel.java [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+
+    @Override
+    public void actionPerformed(AnActionEvent e) {
+      PantsMakeBeforeRun.terminatePantsProcess(e.getProject());
+      PantsConsoleManager.getOrMakeNewConsole(e.getProject()).print(PantsBundle.message("pants.command.terminated"), ConsoleViewContentType.ERROR_OUTPUT);
+    }
+
+    @Override
+    public void update(AnActionEvent event) {
+      // Make the action only clickable when there is active Pants proecess.
```
Comment (baroquebobcat) [inline src/com/twitter/intellij/pants/ui/PantsConsoleViewPanel.java]: nit: s/proecess/process/

Comment (wisechengyi) [inline src/com/twitter/intellij/pants/ui/PantsConsoleViewPanel.java]: corrected. thanks!

**Comment by wisechengyi** [comment]
there's some issue with travis cache causing zinc issue. will likely merge it upon approval.

**Comment by wisechengyi** [comment]
@baroquebobcat test added if you would like to take another look. thanks!

**Pull request merged**